### DISCUSSION TITLE
SupportBrowserHistory.php removes array keys from component property if not present in url query string

### DISCUSSION THREAD

**Discussion by craigkuhns** [livewire/livewire#4077] [7. Bugs (archived)]
Laravel v8.66.0
Livewire v2.7.2

When a public property is an array and that property is in the `protected $queryString` array, keys from that property are removed if they are not present in the url query string at the time the component is initialized. 

For example if I have the following full page component (available at /test) where I have a public `$filters` property and that property is added to the `$queryString` array:

```
<?php

namespace App\Http\Livewire;

use Livewire\Component;

class Test extends Component
{
    public $filters = [
        'search' => '',
        'vendor' => ''
    ];
    protected $queryString = ['filters'];

    public function mount()
    {
        dd($this->filters);
    }

    public function render()
    {
        return view('livewire.test');
    }
}

```

When I visit /test directly in my browser I see the following dumped which is what I'd expect:
```
[
    'search' =>  '',
    'vendor' => ''
]
```

If I visit /test?filters[search]=foo I see:
```
[
    'search' =>  'foo'
]
```

When I'd expect to see:
```
[
    'search' =>  'foo',
    'vendor' => ''
]
```

This leads to undefined index errors in my component and blade files when I assume that `$filters` will have the default key and value pairs in it.

I'm currently working around it by doing the following in `mount()`
```
public function mount()
{
    $defaultFilters = [
        'search' => '',
        'vendor' => ''
    ];     
    $this->filters = array_merge_recursive_distinct($defaultFilters, $this->filters);
}
```
which works but it litters my components with unnecessary duplicate code.

I was able to fix this in Livewire itself by changing line 39 in Features/SupportBrowserHistory.php from
```
$component->$property = $decoded === null ? $fromQueryString : $decoded;
```
to
```
$newPropertyValue = $decoded === null ? $fromQueryString : $decoded;
$component->$property = is_array($component->$property)
    ? array_merge_recursive_distinct($component->$property, $newPropertyValue)
    : $newPropertyValue;
```

*Note:* I'm using a custom helper function array_merge_recursive_distinct to recursively merge the arrays and overwrite duplicate keys with the new value instead of turning them into arrays themselves. For example:
```
$default = [
    'foo' => '',
    'bar' => [
        'baz' => '',
        'mux' => ''
    ]
];

$newValues =  [
    'foo' => 'abc',
    'bar' => [
        'baz' => 'bob'
    ]
]

array_merge_recursive($default, $newValues);
```

returns
```
[
    'foo' => [
        0 => '',
        1 => 'abc'
    ],
    'bar' => [
        'baz' => [
            0 => '',
            1 => 'bob',
        ],
        'mux' => ''
    ]
]

```

Whereas using `array_merge_recursive_distinct` returns
```
[
    'foo' => 'abc',
    'bar' => [
        'baz' => 'bob',
        'mux' => ''
    ]
]
```

I'm willing to work on a pull request with tests to fix this if interested.

**Comment by joshhanley** [comment]
@craigkuhns thanks for reporting the issue. I can see how that would cause problems.

Personally I wouldn't expect it to merge as what is in the querystring, is what should be in the property. But even if we did decide that is how it should work, it would be a breaking change, so we wouldn't be able to merge that in now.

One possible work around would be to support using "except" on keys in an array.

Currently we can do 

```php
public $search = '';

protected $queryString = [
        'search' => ['except' => ''],
];
```

But that only works for individual properties.

I don't think that currently works for keys in an array.

Maybe something like this would be better:

```php
public filters = [];

protected $queryString = [
    'filters' => [
        'search' => ['except' => ''],
        'vendor' => ['except' =>''],
    ]
];
```

Not sure how easy that will be to do, but PR's are welcome if you think that could be useful for you.

Hope this helps!

**Reply by craigkuhns** [reply]
Thanks for the reply Josh. That makes sense about the component property matching what is in the query string. I think what you described would be useful. For now I've extracted my duplicate code out into a trait that handles my use case automatically. I'll try to take a look at what it would take to add your proposed solution because that does make more sense.

I do think there is a bug which is what led me down this path in the first place. If you take my test component and remove the dd from mount so that it renders and then load it in the browser by visiting /test, it updates the url to add the query string. So the url becomes /test?filters[search]=&filters[vendor]=. Which, from my understanding is the expected behavior. 

However, if you visit /test?filters[search]=&filters[vendor]=foo directly, the url changes to /test?filters[vendor]=foo. At that point, the filters property is getting updated to remove the search key from it because that parameter was automatically stripped from the url. To me, it seems like Livewire shouldn't be removing empty query string parameters from the url. Especially when Livewire will add empty query string parameters to the url if the component is loaded at /test. 

I think this issue would be fixed if we're able to get the `protected $queryString = [ 'filters' => [ 'search' => ['except' => ''] ] ];` solution working. However I still think it's unexpected behavior to have parameters removed from the query string when the route is loaded with them explicitly. If that is the case I can try to figure why that is happening and fix that.

**Reply by joshhanley** [reply]
@craigkuhns no worries 🙂

Yeah that seems like it could be a bug. Have a look and doing the except on the array it too hard, then we can at least look into fixing that bug.
